### ISSUE TITLE
Setup automatic code style checker

### ISSUE THREAD

**Issue by nickthetait** [openssl/openssl#5013]
Just like CI and CLA checks, but to verify new PRs fit with desired code style. This task is highly suited to automation and will help improve/standardize the style over time. May also slightly reduce the load that must be taken on in manual code reviews. 

If this is something that is desired, I'm happy to do the research on what tool(s) to use.

**Comment by richsalz** [comment]
This is an interesting idea.  Look at util/openssl-format-source.  We used that to reformat the code, but it then required hand cleanups.  It might be a useful starting point.

**Comment by levitte** [comment]
Note that our coding style isn't always *absolutely* precise in all cases...  which means that an implementation of a style checker needs to allow for this.  The obvious and simple "let's run a code beautifier, make a diff and scream if there's a difference" is probably not the answer for this.

As for static analysis, we do use coverity already...  not that I personally mind another set of checks, just saying that it doesn't have to be highest priority.

**Comment by nickthetait** [comment]
#Questions
1) Does openssl-format-source cover anything in addition to indentation?

2) How is Coverity used currently? I see the data available here https://scan.coverity.com/projects/openssl What I'm not sure about is how anyone notices these findings and converts it into action.

3) Which tool(s) listed below should I try out by actually testing against the OpenSSL codebase?

4) Are there any tools I overlooked and should be included on this list?

# Assumptions
## Requirements
C support
runs on Linux or cloud hosted

## Preferred
no cost
open source
github integration

# Findings
The following static analyzers are in no particular order, except for the first one, Flawfinder. 

## Flawfinder
Flawfinder's simplicity earned it my pick as a great candidate for a secondary SA tool. It should be easy to setup, learn, and derive real benefits from. Additionally this quote is the most important finding in my SA tools research: "No one tool solves all problems, but tools can be a very useful aid in developing secure software. It's generally better to use many tools, and the worst situation is to use no tools."
GPLv2
https://www.dwheeler.com/flawfinder/

## splint
GPLv2
No development since 2010
requires adding annotations to source code
splint.org down (permanently?) https://web.archive.org/web/20171012075322/http://www.splint.org/
instead look at http://lclint.cs.virginia.edu/ & https://github.com/ravenexp/splint
would require resurrection, I spent approximately two hours trying to get it to compile on my system before moving onto other projects.

## scan-build
UIUC
https://clang-analyzer.llvm.org/scan-build.html

## ConQAT
APLv2
No development since 2015
https://www.cqse.eu/en/products/conqat/install/

## Sparse
designed for use on linux kernel, not sure how useful it is to other projects
MIT
https://sparse.wiki.kernel.org/index.php/Main_Page

## Infer
BSD
http://fbinfer.com/
https://github.com/facebook/infer

## CMetrics
GPLv2
No development since 2012
https://github.com/MetricsGrimoire/CMetrics

## CQMetrics
APLv2
https://github.com/dspinellis/cqmetrics

## OCLint
Modified BSD
http://oclint.org/
https://github.com/oclint/oclint

## Frama-C
LGPLv2
http://frama-c.com/

## GCC's warnings
Some plugins are available to make information accessible

# Additional resources on SA tools
## A few excellent lists 
https://samate.nist.gov/index.php/Source_Code_Security_Analyzers.html
https://www.dwheeler.com/essays/static-analysis-tools.html
https://github.com/mre/awesome-static-analysis#cc

## Stand-alone articles
https://dzone.com/articles/10-biggest-mistakes-in-using-static-analysis
https://www.synopsys.com/blogs/software-security/static-analysis-for-security/
https://smackerelofopinion.blogspot.com.au/2017/09/static-analysis-on-linux-kernel.html

**Comment by richsalz** [comment]
To answer question 2:  Coverity mails weekly reports to the openssl-commits mailing list (public list we host) and it's up to people to review the reports and fix things.  We should be more proactive.

**Comment by nickthetait** [comment]
A simple way to raise visibility would be to add a badge to the README. Would you accept a PR if I made one for that? Here's what it looks like for CURL https://github.com/curl/curl/blob/master/README.md

**Comment by richsalz** [comment]
We tend to put that kind of stuff, when we do at all, on our website in the policies section.

**Comment by paulidale** [comment]
I'm not keen on enforcing code formatting.  For new/modified code definitely.  For old code, I'm not so sure.  I'd rather not be force to reformat an entire file when I want to change one line -- the change would risk being lost in the noise.

As for static analysis, I've manually run both [Parfait](http://parfait.us.oracle.com/) and [Clang Static Analyzer](https://clang-analyzer.llvm.org/) over the code.  At some stage I'd like to get these in a CI loop.

**Comment by nickthetait** [comment]
I am suggesting using a formatting tool for new/modified code only. And have it trigger/scan when PRs are submitted.

Happy to do some research/testing/integration to bring SA tools into CI and accessible to everyone. Right now I'm just waiting for direction on which tool to try out first...

The parfait site doesn't resolve for me, it is not a publicly accessible tool right? I can't seem to find much info on it either.

What do you like about using Clang? Would you recommend it?

**Comment by richsalz** [comment]
Our experiences have been that there are too many corner cases for strictly automatic formatting -- at least using the tool we used -- to work 100%.  But if you're interested, please keep plugging away at it.  If nothing else, you might find style policy things that we didn't write down.

Clang and gcc are in an "arms race" for static analysis. :)  Both are improving.

**Comment by kaduk** [comment]
A lot of static analysis tools have a lot of false positives, even coverity and scan-build.  So while I don't have much personal experience with most of the listed tools, it seems that a prerequisite for getting any of them more tightly integrated into the official OpenSSL CI would be someone running them manually and getting a handle for the level of false positives that could be expected.  That would let us get a sense of whether the benefit is worth the noise.

I'm also not entirely clear about what the actionable item(s) are that this issue should remain open for.

**Comment by paulidale** [comment]
Apologies for the Parfait link, it is internal.  The developers are just down the passage from me, which makes for fast answers to questions.  There is nothing particular I like about Clang's analyser.  It seems to do a reasonable job but doesn't seem up to the same level as the dedicated ones -- yet.

I figure that the more of these that are run the better.  At least if the false positives can be controlled.

**Comment by nickthetait** [comment]
Thanks for everyone's input. Here's my recap of feedback so far:
* Current SA tool (coverity) is not sufficiently integrated to the CI toolchain to be significantly useful
* Minimum one tool required, even more preferred
* Little to no preference on which tool to look at first

Actionable items on my part before closing this issue
- [ ] Test out the selected tools, measuring for:
    * Ease of install/config
    * Types of errors that can be detected
    * Noise level -  estimation of how many reported errors are false positives
- [ ] Recommend a SA tool for use by OpenSSL
- [ ] Assist in installation/deployment of the new tool

**Comment by nickthetait** [comment]
I've tried out several different tools so far, but come to an impasse and need a little help. In order to use codecov.io/gcov I need to pass a new flag during compilation. Here is the example: https://github.com/codecov/example-c#usage

I've never dealt with a configure script so complex as OpenSSL's, so I'm not sure the right way to accomplish this. Is modification of this script required, just pass the flag directly like `./config --coverage`, or some other method? How can I check that the coverage flag was actually accepted/used?

**Comment by richsalz** [comment]
ping @levitte

**Comment by levitte** [comment]
The gcc/clang `-coverage` flags takes only one dash, not two.  So, `./config -coverage`

One way to check that it got through like it should is this: `grep '^CFLAGS *=' Makefile`

**Comment by nickthetait** [comment]
Well that was an easy fix, thanks 👍 
Research and testing continues....

**Comment by nickthetait** [comment]
Okay. Nearly ready to report testing results. Will be including some screenshots of the output these tools generate. But I do have a concern about security disclosure. While the SA tools and OpenSSL are publicly available, it is possible that some of the "hits" (aka potential security vulnerabilities) are previously undiscovered problems that will need fixing. So should I censor the function names and line numbers in the screenshots before posting?

**Comment by richsalz** [comment]
No.  Our source is public and anyone could do what you're doing.  If you think you found a security issue, mail it openssl-security.

**Comment by nickthetait** [comment]
Will do.

My first recommendation is to setup a replacement test coverage tool for coverity. The most important feature is producing a "change in coverage" statistic when a PR is submitted. It gives immediate feedback on whether test coverage has grown, shrunk, or no change. Instead of asking people to copy/paste the "yes i updated the tests" when submitting a PR, this would be objectively measured. It would then be up to the maintainers to decide/enforce what tolerances are acceptable. Example of what this could look like: https://github.com/Netflix/security_monkey/pull/967#issuecomment-369249735

Bonus points for:
A) graphing coverage over time - one important metric of the project's technical debt. Here is an example https://coveralls.io/github/Netflix/security_monkey?branch=master
B) showing which files/functions need coverage improved - helps focus test development in the most important or at risk areas

Below are the coverage tools I tested. Several are hosted on sourceforge.net which has been experiencing outages seemingly every time I try to use it over the last two weeks. I've given up waiting on SF, several pages linked below are still not available. :disappointed:

# GCOV
* GPLv3
* https://gcc.gnu.org/onlinedocs/gcc/Gcov.html
* component of GCC, actively maintained
* command line tool producing detailed and verbose output
* most accurate when compiler optimizations are turned off
* tested version 5.4.0

# LCOV
* GPLv2
* http://ltp.sourceforge.net/coverage/lcov.php
* extension that converts GCOV output to HTML
* tested version 1.13
![lcov](https://user-images.githubusercontent.com/11818614/36940352-c00572ac-1efe-11e8-8cfb-3139a1be238a.png)
![lcov2](https://user-images.githubusercontent.com/11818614/36940353-c01ed7ce-1efe-11e8-9861-f17f9df8007d.png)

# GCOVR
* BSD 3-clause
* http://gcovr.com/
* Converts GCOV output to more "compact human-readable summaries"
* Output as text, XML, or HTML
* produces most accurate results when using only one statement per line
* tested version 3.4
![gcovr-summary](https://user-images.githubusercontent.com/11818614/36940360-d7a23bde-1efe-11e8-8f1b-75b2e3dece1c.png)
![gcovr-html](https://user-images.githubusercontent.com/11818614/36940359-d78a7cd8-1efe-11e8-8696-db5ef7503992.png)

# Codecov.io & Coveralls.io
* Two different services but they seem very similar in use
* proprietary but free to open source projects
* Web UI and service to instrument and interpret GCOV output 
* Integrations with many CI and code hosting websites including travis and github
* Example test report against OpenSSL https://codecov.io/gh/nickthetait/openssl, I wasn't able to upload an example test report to coveralls, but I'm sure this is just user error :sweat_smile:
* Both cover A and B bonus features

# CppUnit
* https://www.freedesktop.org/wiki/Software/cppunit/
* LGPL
* designed to work with C++ code, requires some source modification to work with C. For this reason I really didn't consider it to be an option and did not try it out.
* actively maintained by libreoffice

# Xcover
* http://www.xcover.org/ (hosted by sourceforge)

# cunit
* http://cunit.sourceforge.net:80/ss_basic.html

**Comment by richsalz** [comment]
Why do you want to *replace* coverity, as opposed to *augment* it?

**Comment by kaduk** [comment]
Also, I think we already are using coveralls, but the output is not as reliable as one would like due to some sort of quantization or sampling granularity effect.

**Comment by nickthetait** [comment]
@richsalz are my efforts to introduce a new SA tool desired?

@kaduk where can I track down the coveralls output?

**Comment by richsalz** [comment]
More (static) analysis is always useful, yes.

**Comment by nickthetait** [comment]
I guess I was not clear that at this point I'm waiting on a decision from the OpenSSL team on which tool to try out. After that I'll submit PR(s) as necessary to setup and tune it to the project's needs.

**Comment by kaduk** [comment]
@nickthetait My only interaction with the coveralls output is in the hook that runs on pull requests tagged with ```[extended tests]``` and reports the change in coverage percentage.  The coveralls invocation is listed in our ```.travis.yml``` but I am not sure I can link to an example of it right now, since the CI results are not displayed on closed pull requests and I didn't find an open pull request that uses extended tests in some spot checking.  So, I don't know if there's a dashboard page that tracks results over time.

Sorry I can't be more helpful...

**Comment by kroeckx** [comment]
One tool I've used to run is https://github.com/TrustInSoft/tis-interpreter in combinations with the fuzz corpus. I've meant to set it up that this runs on a regular basis, but I haven't found the time yet. This is a tool that does not have any false positives, if it reports an error it's always a real error.

**Comment by kroeckx** [comment]
coveralls results can be seen at https://coveralls.io/github/openssl/openssl

**Comment by kroeckx** [comment]
Not that tis-interpreter is very slow and the longer it runs the slower it gets. So it's not something we want to run on every pull request.

**Comment by mspncp** [comment]
@nickthetait I noticed a while ago that it is possible to sign into AppVeyor and Travis CI using your GitHub login. When you do this, you get your personal CI for free (!). That means whenever you push a branch to nickthetait/openssl, that will automatically kick the CIs for you. So you can play around with the CI _without_ having to create a pull request to openssl/openssl! (This works, because you have a `.travis.yml` and a `appveyor.yml` file in your working copy.)

That means, you can create a branch from master, create a dummy commit containing the text `[extended tests]` and push that branch to your repository. Your private CI will start and execute the extended tests.

I just pushed https://github.com/mspncp/openssl/tree/extended-tests to my repository and this fired up the CI.

I don't know whether you will see the results. At least for AppVeyor it appears that one has to be logged in, because there is no user name in the url:

https://ci.appveyor.com/projects

But maybe you can see my Travis job:

https://travis-ci.org/mspncp/openssl

**Comment by mspncp** [comment]
Look for builds with EXTENDED_TESTS=„yes“, e.g.

https://travis-ci.org/mspncp/openssl/jobs/354430002

**Comment by kroeckx** [comment]
An other tool that I just found and might be useful is https://github.com/kframework/c-semantics, but I really have no idea about if it's any good or not.

**Comment by nickthetait** [comment]
Thanks everyone for your suggestions. Since coveralls.io is already in use, my goal now is to integrate it so a report is posted as part of the "pull-request check" system. Then I'll go back and try some of the suggestions like c-semantics and tis-interpreter.

**Comment by kroeckx** [comment]
I think it's intentional that it's not part of the pull request anymore. If you look at the graph, we don't really have a stable output, so it's giving false positives all the time. As far as I know it now just runs after it's been merged.

**Comment by kroeckx** [comment]
It's run when you enable the extended tests, which also get run after it's merged.

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
